multicore_runtime/wsgi_config.py

```diff
@@ -0,0 +1,208 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Utilities to configure the dispatcher, middleware and environment."""
+
+import logging
+import os
+import threading
+import UserDict
+
+from static_files import static_app_for_regex_and_files
+
+from google.appengine.api import appinfo
+from google.appengine.api import appinfo_includes
+from google.appengine.runtime import wsgi
+
+
+def get_module_config_filename():
+  """Returns the name of the module configuration file (app.yaml)."""
+  module_yaml_path = os.environ['MODULE_YAML_PATH']
+  logging.info('Using module_yaml_path from env: %s', module_yaml_path)
+  return module_yaml_path
+
+
+def get_module_config(filename):
+  """Returns the parsed module config."""
+  with open(filename) as f:
+    return appinfo_includes.Parse(f)
+
+
+def app_for_script(script):
+  """Returns the WSGI app specified in the input string, or None on failure."""
+  try:
+    app, filename, err = wsgi.LoadObject(script)  # pylint: disable=unused-variable
+  except ImportError as e:
+    # Despite nominally returning an error object, LoadObject will sometimes
+    # just result in an exception. Since we're already processing the err
+    # object, we might as well just use that variable to store the exception.
+    err = e
+  if err:
+    # Log the exception but do not reraise.
+    logging.exception('Failed to import %s: %s', script, err)
+    return None
+  else:
+    return app_wrapped_in_user_middleware(app)
+
+
+def app_wrapped_in_user_middleware(app):
+  """Returns the input WSGI app, wrapped in appengine_config middleware."""
+  add_middleware = get_add_middleware_from_appengine_config()
+  if add_middleware:
+    return add_middleware(app)
+  else:
+    return app
+
+
+def get_add_middleware_from_appengine_config():
+  """Tries to import appengine_config and return middleware; fails silently.
+
+  `env_config` is optionally part of GAE-compatible user code.
+  If the user chooses to define the `webapp_add_wsgi_middleware` function
+  there, then we will use it to wrap app scripts. This is used for e.g.
+  appstats. Many user apps do not have or need env_config, so failure
+  to import it or failure to find `webapp_add_wsgi_middleware` is not an
+  error.
+
+  Returns:
+    The appengine_config.webapp_add_wsgi_middleware function or None.
+  """
+  try:
+    import appengine_config  # pylint: disable=g-import-not-at-top
+    try:
+      return appengine_config.webapp_add_wsgi_middleware
+    except AttributeError:
```

Comment:
nitpick: In `cloud_logging.py` you used `getattr(...)` with a default to detect whether the attribute was present or not (for `trace_id`), perhaps for consistency you should do the same here and there?